Automate test masters

chick opened the pull request:
Lots of changes but the main ones of interest at the changes to Readme and the
python scripts in `src/release_scripts`
I don't know if it's necessary to merge this yet, it might be better to perhaps to track this as a `dev` branch
while it is being fleshed out. I have successfully run `build_masters.py` (though I had to force running steps after step 6 because of expected problems with the test)

chick committed:
Working on an automated script to publish snapshots
Tricky stuff is trying to move code to chisel-repo-tools so
chisel-release repo is just data

chick committed:
Moving more stuff to python

chick committed:
cleaning up before tools refactor

chick committed:
Added conditional step running

chick committed:
added getopts

chick committed:
made it through make install

chick committed:
add a few more functions to build_masters

chick committed:
list mode works

chick committed:
use derived function_name instead of manual one

chick committed:
Added some error checking for make test

chick committed:
Added some error checking for make test

chick committed:
Moved dir switching into tools __init__

chick committed:
a bit more cleanup

chick committed:
makefile moved to chisel-repo-tools
build_masters.py has run successfully!

chick committed:
almost forgot to add Makefile

chick committed:
Add some documentation to the README.md

chick committed:
Added checks for verilator, yosys and z3

chick committed:
Added checks for verilator, yosys and z3

ucbjrl commented on scripts/run_check_tests.sh:
```diff
@@ -0,0 +1,34 @@
+
+run_and_check_neg 4.1 grep -n "Tests:.succeeded.*failed.[1-9]" run_tests_test.out
+
+# run_and_check 4.3 tail -100 run_tests_test.out
```
missing newline at EOF?

ucbjrl commented on scripts/setup_env.sh:
```diff
@@ -0,0 +1,9 @@
+source ~/.virtualenvs/cit3/bin/activate
+export PYTHONPATH=/Users/chick/Adept/dev/release-generators/chisel-repo-tools/src
+export VERSIONING=/Users/chick/Adept/dev/release-generators/chisel-repo-tools/src/versioning/versioning.py
+sys.prefix == sys.base_prefix
```
missing newline at EOF

ucbjrl commented on src/repocontributors/repocon.py:
```diff
@@ -62,7 +62,7 @@ def __init__(self, area, reString):
 
 areaREs = [
-        areaRE('documentation', r'doc/'),
+        areaRE('documentation', r'docs/'),
```
I don't know if this is useful in anything other than chisel2. If you aren't actually using it, I'd undo this edit just so it's obvious the file is way out of date.

ucbjrl reviewed (approved):
Great job Chick. I haven't looked at the new python code in any detail, but I like the overall approach.

chick committed:
Delete first try scripts directory
add step counter code to auto generate step numbers
start work on publish_snapshots.py
start trying to get versioning working from within python

chick committed:
remove accidental change to repocon.py

chick commented on src/versioning/versioning.py:
```diff
@@ -691,7 +691,9 @@ def main(argv=None) -> int: # IGNORE:C0111
     program_build_date = str(__updated__)
     program_version_message = '%%(prog)s %s (%s)' % (program_version, program_build_date)
-    program_shortdesc = __import__('__main__').__doc__.split("\n")[1]
+    #program_shortdesc = __import__('__main__').__doc__.split("\n")[1]
```
@ucbjrl I'm not sure what this is doing, my hack is just to get it to run right now.

chick commented:
Current work has been switched to branch `dev` which will soon be PR's to master
